Add tab if alternate login present

ochorocho opened the pull request:
Recently installed Nextcloud for roughly 40 users. Users are supposed to login using Gitlab login provided by sociallogin app.

Not even half of the users noticed the "Alternate Login" button. Therefore i created a tab menu for login. I also introduced a setting in admin area to force alternate login as "active" tab.

<img width="528" alt="Bildschirmfoto 2019-08-22 um 01 04 16" src="https://user-images.githubusercontent.com/4623070/63474259-c878f280-c478-11e9-9e8e-25aa12a2b99d.png">

When no alternate login is available:

<img width="384" alt="Bildschirmfoto 2019-08-24 um 19 47 40" src="https://user-images.githubusercontent.com/4623070/63641586-052f2e80-c6b1-11e9-9835-128b615fd252.png">

rullzer commented:
Could you rebase your PR ;)
As 12010 files changes seems a tad much :wink:

ochorocho commented:
> Could you rebase your PR ;)
> As 12010 files changes seems a tad much 😉

Tbh, was late yesterday and i screwed it up while signing my commits...

@rullzer do you like the idea?

You can test it here: https://nextcloud.knallimall.org/

rullzer commented:
I'm fine but I'd like the JS experts and designers to have a look ;)

kesselb commented on core/Controller/LoginController.php:
```diff
@@ -189,6 +189,8 @@ public function showLoginForm(string $user = null, string $redirect_url = null):
 		$parameters = [
 			'alt_login' => OC_App::getAlternativeLogIns(),
+			'alt_login_tab_force' => (\OC::$server->getConfig()->getAppValue('core', 'alternate_login_force', 'no') === 'no') ? false : true,
+
```
```suggestion
```

ochorocho replied:
Remove empty line

kesselb commented on core/Controller/LoginController.php:
```diff
@@ -189,6 +189,8 @@ public function showLoginForm(string $user = null, string $redirect_url = null):
 
 		$parameters = [
 			'alt_login' => OC_App::getAlternativeLogIns(),
+			'alt_login_tab_force' => (\OC::$server->getConfig()->getAppValue('core', 'alternate_login_force', 'no') === 'no') ? false : true,
```
```suggestion
			'alt_login_tab_force' => \OC::$server->getConfig()->getAppValue('core', 'alternate_login_force', 'no') === 'yes',
```

ochorocho replied:
Adapted the change to the value set in `config.php`

kesselb commented on settings/templates/settings/admin/server.php:
```diff
@@ -26,6 +26,15 @@
 
 ?>
 
+<div class="section" id="shareAPI">
```
Please pick a different id.

ochorocho replied:
Removed the code due to moving setting to `config.php`

kesselb commented on settings/Settings/Admin/Server.php:
```diff
@@ -81,6 +81,7 @@ public function getForm() {
 			'cronErrors'		  => $this->config->getAppValue('core', 'cronErrors'),
 			'cli_based_cron_possible' => function_exists('posix_getpwuid'),
 			'cli_based_cron_user' => function_exists('posix_getpwuid') ? posix_getpwuid(fileowner(\OC::$configDir . 'config.php'))['name'] : '',
+			'alternate_login_force'		  => $this->config->getAppValue('core', 'alternate_login_force'),
```
Hmm. Not sure about this. I would rather put this config flag into `config.php` instead the db. cc @rullzer

ochorocho replied:
Makes sense. I moved this option to `config.php`. I wasn't happy with that either.

kesselb commented on core/templates/login.php:
```diff
@@ -1,17 +1,36 @@
-<?php }
+<ul id="login-tabs" class="login-tabs">
+    <li class="tabs">
+        <div class="title <?php echo $_['alt_login_tab_force'] ?: 'active' ?>" data-tab="tab-login">
```
If there are no alternative logins there is always a tab standard right? Would you mind to add a screenshot for this case? I think the tabs should be hidden if there is only the default login.

ochorocho replied:
Yes, you are right. Tabs are now hidden if there is only the default login.

<img width="384" alt="Bildschirmfoto 2019-08-24 um 19 47 40" src="https://user-images.githubusercontent.com/4623070/63641041-1b84bc80-c6a8-11e9-8218-dd87d276dca5.png">

kesselb reviewed (commented):
I'm not a fan. I think we should improve the visibility of the alternative logins instead of adding a new visual element (which is not used anywhere else in nextcloud). For example the stackoverflows login screen looks very clean.
![image](https://user-images.githubusercontent.com/3902676/63638917-4eba5200-c68e-11e9-8f1e-aaefe31104a9.png)

splitt3r commented:
Off topic: @ochorocho Awesome that you contribute to GitLab and now Nextcloud :heart_eyes: Great work!

jancborchardt commented:
Good initiative @ochorocho!

Agree with @kesselb here that instead of making the design more complex, the other logins should be made more present by being moved up like on other platforms. :)

Then not even an extra admin setting to set the active tab is necessary because there are no tabs. Easier for everyone. :)

ochorocho commented:
Ok :-) Still like the tab view thing.
Created a little theme which does what i need, see https://github.com/ochorocho/nextcloud-theme-tablogin

Works with NC 16 and NC 17 (current master)

jancborchardt commented:
@ochorocho @kesselb would someone of you be up to change it as @kesselb described? Would certainly be good to fix this in core and not just as a theme. :)

kesselb commented:
Adding a logo and another color seems to be more complicated. I think this information should be provided by https://github.com/zorn-v/nextcloud-social-login and passed as additional data (e.g. a logo) to https://github.com/nextcloud/server/blob/810ee7d811c4cf0291add770cb89e5b6d11104b9/lib/private/legacy/app.php#L648-L650 

As a start we should set a custom class to `<li>` based on the name or href to apply custom styles more easily.

https://github.com/nextcloud/server/blob/64c4bb5bce52599d9df4007ba9d622840fee1213/core/templates/login.php#L11-L13

ochorocho committed:
Move alternative login above standard login, optimize alignment

Signed-off-by: ochorocho <rothjochen@gmail.com>

ochorocho commented:
@jancborchardt just moved the buttons above the form and optimized alignment of the buttons

See https://nextcloud.knallimall.org/

kesselb commented:
![image](https://user-images.githubusercontent.com/3902676/63712588-43a62400-c83e-11e9-845f-bbec85918e09.png)

I would suggest to add "log in with" to make the purpose of the button more clear.

kesselb commented:
https://github.com/zorn-v/nextcloud-social-login/blob/master/css/style.css Hmm. There is already a style sheet with logos for github.

ochorocho commented:
Yes, only for the pre-defined oauth provider but not for "OpenID Connect/Custom OpenID Connect"

<img width="675" alt="Bildschirmfoto 2019-08-26 um 20 42 59" src="https://user-images.githubusercontent.com/4623070/63714580-5ae71080-c842-11e9-9dcd-8615934e9436.png">

jancborchardt commented:
Great stuff @ochorocho!
Agree with adding the "Log in with …" before the provider that @kesselb mentioned. Also, right now it’s a bit confusing cause there are 3 primary buttons:
- [ ] The buttons should be colored according to the provider. Google red, Twitter light blue, Facebook dark blue, GitLab dark purple, GitHub dark grey etc.
- [ ] Also I would add icons (monochrome, that is only white) to the left inside the icons for better recognition, as this is also standard for "Log in with" buttons. We have some of these icons already in https://github.com/nextcloud/server/tree/master/core/img (like twitter.svg)
- [ ] There should be a tiny bit more whitespace between the log in buttons and the form for username and password, just for visual separation.

Then I think it would be super.

ochorocho commented:
@jancborchardt @kesselb this is what i'm going to do.


Todo Nextcloud
* More space between login buttons

Todo in Social login
* Color Buttons
* Add predifined icons (monochrom)
* Add "Login in with"

jancborchardt commented:
Yup, sounds great @ochorocho! :) Thanks a lot.

And maybe with the visual changes to the buttons it’s also fine to put them below the "Log in" button again. So like:

```
[       Username      ]
[       Password      ]
(       Log in →      )
          or
( Log in with Google  )
( Log in with GitHub  )
```

ochorocho committed:
Add more space to buttonsy

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho committed:
Add more space between form and alternate login

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho commented:
This is how it will like now...
Still need open a pr for the sociallogin repo.

![Screenshot_2019-09-03 Nextcloud(1)](https://user-images.githubusercontent.com/4623070/64207827-b39b5680-ce9d-11e9-8446-e6b29ec2fb58.png)
![Screenshot_2019-09-03 Nextcloud](https://user-images.githubusercontent.com/4623070/64207828-b433ed00-ce9d-11e9-96da-215697c3a5fb.png)

skjnldsv commented:
I guess no one will have that many, right? ^^'

ochorocho commented:
> I guess no one will have that many, right? ^^'

@skjnldsv i depends on how much alternative logins are configured. In my case i'm using `sociallogin` and i could have endless providers configure, currently only using 1 on my production install. :-)

ochorocho committed:
Add default style for alternative login

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho commented:
Created a PR for the `sociallogin` app

https://github.com/zorn-v/nextcloud-social-login/pull/109

jancborchardt reviewed (approved):
@ochorocho this looks perfect now, good stuff! Thanks a lot for the modifications :)

zorn-v commented:
I think it is bad idea to add `<?php p($l->t('Log in with')) ?>` in nextcloud template, it should be done on socialligin app side. 
Not only social login use "alternate logins". Thre is registration app for example (https://apps.nextcloud.com/apps/registration) 
"Log in with Register" - seriosly ? :smile:

zorn-v commented on core/templates/login.php:
```diff
@@ -4,14 +4,14 @@
-<?php }
+    <div id="alternative-logins">
+        <?php foreach($_['alt_login'] as $login): ?>
+            <a class="button primary <?php p($login['style']); ?>" href="<?php print_unescaped($login['href']); ?>" >
```
$login['style'] is not defined. It shold be `p(isset($login['style']) ? $login['style'] : '')` to prevent "undefined index 'style'" warnings in log

skjnldsv replied:
Yeah, I'd say let the other apps register the entire string. @jancborchardt ?

ochorocho replied:
> I think it is bad idea to add `<?php p($l->t('Log in with')) ?>` in nextcloud template, it should be done on socialligin app side.
> Not only social login use "alternate logins". Thre is registration app for example (https://apps.nextcloud.com/apps/registration)
> "Log in with Register" - seriosly ? 😄

Yep, makes sense. i'll move that to the sociallogin app.

ochorocho replied:
Now `sociallogin` app registers the entire string.

ochorocho commented:
> I think it is bad idea to add `<?php p($l->t('Log in with')) ?>` in nextcloud template, it should be done on socialligin app side.
> Not only social login use "alternate logins". Thre is registration app for example (https://apps.nextcloud.com/apps/registration)
> "Log in with Register" - seriosly ? 😄

Yep, makes sense. i'll move that to the sociallogin app. :-)

ochorocho committed:
Move "Log in with" label to "sociallogin" app

https://github.com/nextcloud/server/pull/16832#issuecomment-528197965

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho committed:
Prevent undefined for $login['style']

https://github.com/nextcloud/server/pull/16832#discussion_r321075261

Signed-off-by: Jochen <rothjochen@gmail.com>

zorn-v commented:
Need some style changes. If you have long names, buttons should be centered I think

![изображение](https://user-images.githubusercontent.com/12619075/64395682-6d611580-d09e-11e9-9ebc-5479a4491aa2.png)

zorn-v commented:
Also I think to bring back `width: 10px` for `::after`. Text in buttons without icons is slip to the right

zorn-v commented:
I play around with `guest.css` styles
- remove `width` in `.wrapper` and `form`
- remove `text-align: left;` in `form fieldset`
- add `width: 260px; margin: 0 auto;` to `#submit-wrapper`

and got
![изображение](https://user-images.githubusercontent.com/12619075/64400375-5ecf2a00-d0af-11e9-963f-f4682507d3a0.png)

ochorocho commented:
@zorn-v hm... makes sense. I'll try to make this more flexible and at the same time if names are shorter i'll try to keep a min-width of the "Log in" Button.

zorn-v commented:
I already add "guest.css" in social login app with "!important" overrides, but it is bad practice I think.

jancborchardt commented:
Good on removing the white border for the colored buttons!

Some feedback:
- Only the "Log in" button should have class primary, the others should not be blue.
- Make sure to test on mobile where the text in the buttons needs to ellipsize at some point and shouldn’t break out of the viewport.

ochorocho commented:
imho, this should be handled by the core i think as it should take the registration app and such into account

ochorocho commented:
>     * Only the "Log in" button should have class primary, the others should not be blue.

What color would you suggest as "not blue" ? :-) 

>     * Make sure to test on mobile where the text in the buttons needs to ellipsize at some point and shouldn’t break out of the viewport.

Currently thinking about using min-width/max-width. And using css ellipsis to shorten text if it does not fit in for mobile devices

zorn-v commented:
> And using css ellipsis to shorten text if it does not fit in for mobile devices

And you can get many buttons with "Log in with (translate is very long) ..." :wink:

ochorocho commented:
@zorn-v Yep, another idea would be to remove "Log in with" on small viewports.

zorn-v commented:
I think it is good compromise

zorn-v commented:
> What color would you suggest as "not blue" ? :-)

I think someting like `border: 0` for colored buttons. Blue buttons for alternative logins without borders is ugly I think.

jancborchardt commented:
> What color would you suggest as "not blue" ? :-)

Simply the `class="primary"` needs to be removed from the button, then it should work. :)

ochorocho committed:
Optimize responsiveness, remove primary class

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho commented:
@jancborchardt @zorn-v 

* removed class `primary` , after that registration buttone was white background with white text, so i changed the text-color to blue. 
* Removed "Log in with" on small viewports (< 1024px)
* `min-width` was set to align width to "Log in"-Button, if more space needed it will take up to 400px

![responsiveness](https://user-images.githubusercontent.com/4623070/64449470-488d9200-d0e0-11e9-9fd0-34e682536ea6.gif)

![Screenshot_2019-09-06 Nextcloud](https://user-images.githubusercontent.com/4623070/64449835-13357400-d0e1-11e9-94e2-6ed371db10b2.png)

jancborchardt reviewed (approved):
Super impressive work @ochorocho! Looks perfect now :rocket: 

Please review too @nextcloud/designers :)

jancborchardt commented:
Oh just saw the:

> Removed "Log in with" on small viewports (< 1024px)

I would say this makes it confusing. I’d rather say cut the icon on mobile, or ellipsize. But the label needs to be explanatory. Otherwise it’s like "Google? What, is that a link to the service?"

(also small typo: It’s "Salesforce" not "Saleforce", not sure in which component that is. :)

ochorocho committed:
Hide only icons for buttons

https://github.com/nextcloud/server/pull/16832#issuecomment-529115417

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho commented:
@jancborchardt only icons hidden on smal devices now...

![hide_icons_only](https://user-images.githubusercontent.com/4623070/64480389-1efb6600-d1c7-11e9-9611-b1159ed5ea0e.gif)

kesselb commented:
Thank you @ochorocho for this great contribution :tada: and also thanks to @zorn-v for your feedback over here :+1:

zorn-v commented on core/templates/login.php:
```diff
@@ -4,14 +4,13 @@
+    <div id="alternative-logins">
+        <?php foreach($_['alt_login'] as $login): ?>
+            <a class="button <?php p(isset($login['style']) ? $login['style'] : ''); ?>" href="<?php print_unescaped($login['href']); ?>" >
+                <?php echo $login['name']; ?>
```
`echo` should not be used in templates

ochorocho replied:
Yep, a left over from develpment will correct this. Thank you.

ochorocho replied:
Fixed now

skjnldsv commented on core/templates/login.php:
```diff
@@ -4,14 +4,13 @@
+    <div id="alternative-logins">
+        <?php foreach($_['alt_login'] as $login): ?>
+            <a class="button <?php p(isset($login['style']) ? $login['style'] : ''); ?>" href="<?php print_unescaped($login['href']); ?>" >
+                <?php echo $login['name']; ?>
```
```suggestion
                <?php p($login['name']); ?>
```

ochorocho replied:
Already mentioned by zorn. I'll correct this later today.
Thank you. :-)

ochorocho replied:
using `<?php p($login['name']); ?>` now.

skjnldsv commented on core/templates/login.php:
```diff
@@ -4,14 +4,13 @@
-<?php }
+    <div id="alternative-logins">
+        <?php foreach($_['alt_login'] as $login): ?>
+            <a class="button <?php p(isset($login['style']) ? $login['style'] : ''); ?>" href="<?php print_unescaped($login['href']); ?>" >
```
Also, if you could add `aria-label` properties for those links :)
Like
```html
<a aria-label="<?php p($login['label']); ?>"
```

And maybe pass text like "Alternate login with Google" ?
@jancborchardt

ochorocho replied:
Yep, makes sense. I would go with "Log in with Google" as we already use this as button label.

skjnldsv replied:
I would emphasis the fact that it's an alternate option, but it's fine too I guess :)

ochorocho replied:
True :-) .... so i'll add another label with "aria-label" to the array of "alternative logins" as i did with "style"

ochorocho replied:
Now using aria-label "Alternate login with Google", each app can handle `aria-label`iteself.
:-)

jancborchardt replied:
@skjnldsv @ochorocho the `aria-label` is actually not needed here and in fact will just duplicate the content of the button itself. :) The text of the button is already "Log in with Google" so that’s perfect, no separate stuff needed for accessibility.

Or am I missing something?

skjnldsv replied:
Ah you're right! I confused myself with something else! The aria-label if for text that is not visible. We can indeed remove it! 

Sorry @ochorocho

zorn-v replied:
Guys, is you don't missing the fact that translations of "Login with" may be  longer that min width of the button ? And "icons" may be only for predefined styles.

I'm thinking about adding "dynamic styles" and something like "icon url" in social ligin provider settings...

ochorocho replied:
So you want to have an "upload icon"-Button  and "define button color" using a color selector in sociallogin app.

This requires...
* 2 additional fields in sociallogin app settings
* Extend core template to use this icons and color

kesselb replied:
Sounds like a good enhancement for another pull request. I guess for now it should be enough to remove the aria label again ;)

ochorocho replied:
Removed the aria label.

jancborchardt replied:
> I'm thinking about adding "dynamic styles" and something like "icon url" in social login provider settings...

Let’s please keep it simple. The wording should be consistent and every service has a defined color. Should should only be able to say if it’s possible to log in with a service or not, but Nextcloud admins are not designers or copywriters.

ochorocho committed:
Add aria label

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho commented:
I thinks this is ready now :-)

kesselb commented on core/templates/login.php:
```diff
@@ -4,14 +4,13 @@
-<?php }
+    <div id="alternative-logins">
+        <?php foreach($_['alt_login'] as $login): ?>
+            <a class="button <?php p(isset($login['style']) ? $login['style'] : ''); ?>" aria-label="<?php p($login['label']); ?>" href="<?php print_unescaped($login['href']); ?>" >
```
```suggestion
            <a class="button <?php p($login['style'] ?? ''); ?>" aria-label="<?php p($login['label'] ?? ''); ?>" href="<?php print_unescaped($login['href']); ?>" >
```

1) null coalesce makes it a bit more readable
2) label looks like a new attribute to me. we should use null coalesce here as well.

ochorocho replied:
1. yep, changed it
2. yep, its a new attribute. Added `$login['name']` as fallback

kesselb replied:
Thx

zorn-v replied:
Again, `label` may be (and will be) undefined.

ochorocho committed:
Add null coalesce for label and style attribute

https://github.com/nextcloud/server/pull/16832#discussion_r322380822

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho committed:
Added name as fallback, in case label is not set

https://github.com/nextcloud/server/pull/16832#discussion_r322380822

Signed-off-by: Jochen <rothjochen@gmail.com>

ochorocho committed:
Remove aria label

https://github.com/nextcloud/server/pull/16832#discussion_r322078592

Signed-off-by: Jochen <rothjochen@gmail.com>

skjnldsv commented on core/css/guest.css:
```diff
@@ -95,21 +95,21 @@ body {
+
 .wrapper {
-	width: 300px;
+	max-width: 100%;
```
This could lead to issues in other layouts, but we'll have a look later I guess, since we're at the start of 18!

skjnldsv commented:
A few unrelated failures :)
Merging!
